capsules/extra/src/nmea_i2c_mt.rs

```diff
@@ -0,0 +1,238 @@
+use kernel::ErrorCode;
+
+// We can only read a small-ish number of bytes at a time
+// otherwise the GNSS device will crash and hold the I2C
+// clock low. 32 is the maximum we can read
+pub const I2C_BUFFER_LEN: usize = 24;
+
+pub const NMEA_BUFFER_LEN: usize = 128;
+
+pub struct I2cNmea<'a, I: I2CDevice> {
+    sentence_buffer: TakeCell<'static, [u8]>,
+    i2c_buffer: TakeCell<'static, [u8]>,
+    nmea_offset: Cell<usize>,
+    i2c: &'a I,
+    client: OptionalCell<&'a dyn NmeaClient>,
+    write: Cell<(bool, usize)>,
+}
+
+impl<'a, I: I2CDevice> I2cNmea<'a, I> {
+    pub fn new(i2c: &'a I, i2c_buffer: &'static mut [u8]) -> Self {
+        I2cNmea {
+            sentence_buffer: TakeCell::empty(),
+            i2c_buffer: TakeCell::new(i2c_buffer),
+            nmea_offset: Cell::new(0),
+            i2c,
+            client: OptionalCell::empty(),
+            write: Cell::new((false, 0)),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> NmeaDevice<'a> for I2cNmea<'a, I> {
+    fn set_client(&self, client: &'a dyn NmeaClient) {
+        self.client.set(client);
+    }
+
+    fn read_sentence(
+        &self,
+        sentence: &'static mut [u8],
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        match self.i2c_buffer.take() {
+            Some(buffer) => {
+                let i2c_buf_len = buffer.len();
+                self.write.set((false, 0));
+
+                match self.i2c.read(buffer, i2c_buf_len) {
+                    Ok(()) => {
+                        self.sentence_buffer.replace(sentence);
+                        Ok(())
+                    }
+                    Err((e, buf)) => {
+                        self.i2c_buffer.replace(buf);
+                        Err((e.into(), sentence))
+                    }
+                }
+            }
+            None => Err((ErrorCode::BUSY, sentence)),
+        }
+    }
+
+    fn write_sentence(
+        &self,
+        buffer: &'static mut [u8],
+        length: usize,
+    ) -> Result<(), (ErrorCode, &'static mut [u8])> {
+        self.write.set((true, length));
+
+        match self.i2c.write(buffer, length) {
+            Ok(()) => Ok(()),
+            Err((e, buf)) => Err((e.into(), buf)),
+        }
+    }
+}
+
+impl<'a, I: I2CDevice> I2CClient for I2cNmea<'a, I> {
+    fn command_complete(&self, buffer: &'static mut [u8], status: Result<(), i2c::Error>) {
+        if self.write.get().0 {
+            if let Err(e) = status {
+                self.client.map(|call| {
+                    call.callback(buffer, 0, Err(e.into()));
+                });
+            } else {
+                self.client.map(|call| {
+                    call.callback(buffer, self.write.get().1, Ok(()));
+                });
+            }
+        } else {
+            self.sentence_buffer.take().map(|nmea_buf| {
+                let i2c_buf_len = buffer.len();
+
+                if let Err(e) = status {
+                    self.i2c_buffer.replace(buffer);
+
```

Comment:
Set `nmea_offset` to zero? If there was an I2C error, we've likely lost some data, so we need to start over.